sklearn/linear_model/_glm/glm.py

```diff
@@ -20,13 +24,746 @@
+        sqrt_h = np.sqrt(self.h)
+
+        b = np.r_[-self.g / sqrt_h, -self.sqrt_P * self.coef[:n_features]]
+
+        if self.linear_loss.fit_intercept:
+            n_dof = n_features + 1
+
+            def matvec(x):
+                # A @ x with intercept
+                # We assume self.sqrt_P to be 1-d array of shape (n_features,),
+                # representing a diagonal matrix.
+                return np.r_[sqrt_h * (self.X @ x[:-1] + x[-1]), self.sqrt_P * x[:-1]]
+
+            def rmatvec(x):
+                # A.T @ x with intercept
+                return np.r_[
+                    self.X.T @ (sqrt_h * x[:n_samples]) + self.sqrt_P * x[n_samples:],
+                    sqrt_h @ x[:n_samples],
+                ]
+
+        else:
+            n_dof = n_features
+
+            def matvec(x):
+                # A @ x without intercept
+                return np.r_[sqrt_h * (self.X @ x), self.sqrt_P * x]
+
+            def rmatvec(x):
+                # A.T @ x without intercept
+                return self.X.T @ (sqrt_h * x[:n_samples]) + self.sqrt_P * x[n_samples:]
+
+        # Note that initializing LinearOperator seems to have some surprisingly sizable
+        # overhead.
+        A = scipy.sparse.linalg.LinearOperator(
+            shape=(n_samples + n_features, n_dof), matvec=matvec, rmatvec=rmatvec
+        )
+        # Note that the choice of atol is essential for stability and for computation
+        # time. For n_samples > n_features, we most certainly have a least squares
+        # problem (no solution to the linear equation A x = b), such that the following
+        # stopping criterion with residual r = b - A x applies:
+        #   ||A' r|| <= atol * ||A|| * ||r||.
+        # As we get the Frobenius norm of A and the norm of r, ||A|| and ||r||
+        # respectively, for free by LSMR, we use it to set a tighter atol by dividing
+        # by ||A|| * min(1, ||r||) which is very likely larger than 1. The effective
+        # stopping criterion becomes approximately
+        #   ||A' r|| <= self.tol * ||r||   with   x = coef_newton   and
+        #   1/2 * ||r||^2 = 1/2 * x X' diag(h) X x + g X x
+        #                 + 1/2 * l2_reg_strength * (x+x0) P (x+x0)
+        #                 + 1/2 *||g/sqrt(h)||^2
+        # Note that coef=x+x0 is just the solution after an iteration. In particular
+        # note that this is just the Taylor series of the objective with the zero
+        # order term obj(x=0) replaced by 1/2 * ||g/sqrt(h)||^2.
+        result = scipy.sparse.linalg.lsmr(
+            A,
+            b,
+            damp=0,
+            atol=self.tol / (self.A_norm * max(1, self.r_norm)),
+            btol=self.tol,
+            show=self.verbose >= 3,
+        )
+        self.coef_newton = result[0]
+        # Estimated Frobenius norm of A and norm of residual r for tolerance of
+        # next iteration.
+        self.A_norm = result[5]
+        self.r_norm = result[3]
```

Comment:
We might want to issue a warning (and suggest increasing the regularization) if `result[1] == 7` (`istop`) which means that `lsmr` has not converged with the default maxiter` (which should be `n_features` by construction in our case), possibly because of ill-conditioning.

Instead of warning might even want to experiment with doing more iterations of `lsmr` when we detect this situation according to the docstring of `lsmr`, see: https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_isolve/lsmr.py

and then only warn if we do not converge with `maxiter=2 * n_features` in total for instance (e.g. running `lsmr` twice in a row if the first call resulted in `istop == 7`).

Note that issuing warning when the inner solver failed to converge while still proceeding with the outer loop might just result in a wall of warnings and a badly converged model. Maybe we could also choose to just early interrupt the outer solver loop with a `ConvergenceWarning` in such a case. This is related to: https://github.com/lorentzenchr/scikit-learn/pull/6